src/automaton.cpp

```diff
@@ -0,0 +1,484 @@
+	X(Seq, KEYWORD_SEQ, "seq")                                                 \
+	X(Tuple, KEYWORD_TUPLE, "tuple")                                           \
+	X(Struct, KEYWORD_STRUCT, "struct")                                        \
+	X(Union, KEYWORD_UNION, "union")
+
+#define X(name, token_tag, string) name,
+enum Values { Error, END_STATES Count };
+#undef X
+#define X(name, token_tag, string) #name,
+constexpr char const* end_states[] = { "Error", END_STATES };
+#undef X
+}
+#define X(name, token_tag, string) TokenTag::token_tag,
+constexpr TokenTag token_tags[] = { END_STATES };
+#undef X
+#define X(name, token_tag, string) string,
+static InternedString fixed_strings[] = { END_STATES };
+#undef X
+#undef END_STATES
+
+constexpr Automaton make() {
+#define new_state() (--state_counter)
+	Automaton a{};
+
+	int state_counter = state_count;
+
+	int start_state = new_state();
+
+	add_string(a, state_counter, EndStates::If, "if");
+	add_string(a, state_counter, EndStates::For, "for");
+	add_string(a, state_counter, EndStates::Else, "else");
+	add_string(a, state_counter, EndStates::Fn, "fn");
+	add_string(a, state_counter, EndStates::Then, "then");
+	add_string(a, state_counter, EndStates::Return, "return");
+	add_string(a, state_counter, EndStates::While, "while");
+	add_string(a, state_counter, EndStates::Match, "match");
+	add_string(a, state_counter, EndStates::True, "true");
+	add_string(a, state_counter, EndStates::False, "false");
+	add_string(a, state_counter, EndStates::Array, "array");
+	add_string(a, state_counter, EndStates::Null, "null");
+	add_string(a, state_counter, EndStates::Seq, "seq");
+	add_string(a, state_counter, EndStates::Tuple, "tuple");
+	add_string(a, state_counter, EndStates::Struct, "struct");
+	add_string(a, state_counter, EndStates::Union, "union");
+
+	return a;
+#undef new_state
+}
+
+}
+
+static void print_error(char const* p) {
+	printf("Error -- last two chars are: %c%c\n", *(p-2), *(p-1));
+}
+
+static void push_identifier_or_keyword(Automaton const& a, TokenArray& ta, string_view str) {
+
+	int state = state_count - 1;
+	for (int i = 0; i < str.size(); ++i) {
+		state = a.go(state, str.begin()[i]);
+		if (state == KeywordLexer::EndStates::Error)
+			break;
+	}
+
+	if (KeywordLexer::EndStates::Count <= state || state == KeywordLexer::EndStates::Error) {
+		ta.push_back({
+			TokenTag::IDENTIFIER,
+			InternedString(str.begin(), str.size()),
+			0, 0, 0, 0
+		});
+	} else {
+		ta.push_back({
+			KeywordLexer::token_tags[state - 1],
+			KeywordLexer::fixed_strings[state - 1],
+			0, 0, 0, 0,
+		});
+	}
+}
+
+TokenArray tokenize(char const* p) {
+	// Implementation detail: We store indices into the source buffer
+	// in the source_locations at first, then resolve them in another pass.
+
+	char const* const code_start = p;
+
+	constexpr Automaton a = MainLexer::make();
+	constexpr Automaton ka = KeywordLexer::make();
+
+	TokenArray ta;
+	while (*p == ' ' || *p == '\t' || *p == '\n') ++p;
```

Comment:
Used twice, low level and hard to look at. Free macro / lambda

Reply:
got it